### PR TITLE
Add warm up in evaluation_loop to get more accurate performance data

### PR DESCRIPTION
**Pull request by zxd1997066**
Add warm up in evaluation_loop, this will make performance metrics more accurate when using torch_compile, take `text-classification+albert-base-v1` as example:
`python -u ./transformers/examples/pytorch/text-classification/run_glue.py --model_name_or_path albert-base-v1 --task_name MRPC --do_eval --max_seq_length 16 --learning_rate 2e-5 --overwrite_output_dir --output_dir /tmp/tmp_huggingface/ --torch_compile --report_to=none`
before this PR:
```
***** Running Evaluation *****
[INFO|trainer.py:3831] 2024-08-12 07:22:17,390 >>   Num examples = 408
[INFO|trainer.py:3834] 2024-08-12 07:22:17,390 >>   Batch size = 8
100%|████████████████████████████████████████████████████████████████████████| 51/51 [00:00<00:00, 69.95it/s]
***** eval metrics *****
  eval_accuracy               =     0.5637
  eval_combined_score         =     0.6024
  eval_f1                     =     0.6411
  eval_loss                   =     0.6855
  eval_model_preparation_time =     0.6604
  eval_runtime                = 0:00:06.56
  eval_samples                =        408
  eval_samples_per_second     =     62.169
  eval_steps_per_second       =      7.771
```
after this PR:
```
***** Running Evaluation *****
[INFO|trainer.py:3831] 2024-08-12 07:11:18,818 >>   Num examples = 408
[INFO|trainer.py:3834] 2024-08-12 07:11:18,818 >>   Batch size = 8
100%|██████████| 51/51 [00:00<00:00, 68.18it/s]
***** eval metrics *****
  eval_accuracy               =     0.5637
  eval_combined_score         =     0.6024
  eval_f1                     =     0.6411
  eval_loss                   =     0.6855
  eval_model_preparation_time =     6.4489
  eval_runtime                = 0:00:00.76
  eval_samples                =        408
  eval_samples_per_second     =    533.526
  eval_steps_per_second       =     66.691
```

### CODE REVIEW

**Comment by amyeroberts** [comment]
Hi @zxd1997066, thanks for opening a PR! 

How "accurate" the measurement is here depends on all sorts of situations -- for people not compiling the model this add the overhead of a forward pass, which isn't representative of preparation time -- and so adding would both be misleading for some users and would be a breaking change as users reported metrics to date would suddenly change. 

If this is something you want to include - subclassing trainer and overriding the evaluation loop is a quick way to incorporate into your own evals!

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -3800,6 +3800,9 @@ def evaluation_loop(
                 if self.is_deepspeed_enabled
                 else self.accelerator.prepare_model(model, evaluation_mode=True)
             )
+            example_inputs = dict(next(iter(dataloader)))
```
Comment (chuanqi129) [inline src/transformers/trainer.py:3803]: Can we add some check for accelerator mode in here? E.g. if the torch.compile enabled, we go into this warm up stage to exclude the compilation time in the performance statistic

Comment (amyeroberts) [reply src/transformers/trainer.py:3803]: You can exclude compilation time by passing in an already compiled model to the trainer. One can also chose to do a warm-up step before passing to the trainer

Comment (chuanqi129) [reply src/transformers/trainer.py:3803]: Hi @amyeroberts, thanks for your comment. Trainer has a parameter named `--torch_compile`, and it will call `accelerate` package in trainer internal to use `torch.compile()` to mark the model. But the compilation occurs in the model first run. So we need warm-up to exclude the compilation time. Previously, we want to warm-up/compile the model in accelerate internally, refer https://github.com/huggingface/accelerate/pull/2964/files, but we just need to apply `with torch.no_grad()` context in inference mode, accelerate can't know the model is for inference or training. That's why we open this PR.

Comment (amyeroberts) [reply src/transformers/trainer.py:3803]: @chuanqi129 I understand, but this approach is far too brute-force to be something we would marge in: suddenly all users would see a large increase in init time because of this initialization, even if their model isn't compiled. As I mentioned before, you can run a custom loop which would pass in a pre-compiled + warmed-up model or override the evaluation loop for your own needs.
